src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DefaultRazorComponentSearchEngine.cs

```diff
@@ -0,0 +1,93 @@
+        /// Implicitly, this method inherits any assumptions made by TrySplitNamespaceAndType.
+        /// </remarks>
+        /// <param name="tagHelper">A TagHelperDescriptor to find the corresponding Razor component for.</param>
+        /// <returns>The corresponding DocumentSnapshot if found, null otherwise.</returns>
+        public override async Task<DocumentSnapshot> TryLocateComponentAsync(TagHelperDescriptor tagHelper)
+        {
+            if (tagHelper is null)
+            {
+                return null;
+            }
+
+            DefaultRazorTagHelperBinderPhase.ComponentDirectiveVisitor.TrySplitNamespaceAndType(tagHelper.Name, out _, out var typeSpan);
+            var typeName = tagHelper.Name.Substring(typeSpan.Start, typeSpan.Length);
+
+            foreach (var project in _projectSnapshotManager.Projects)
+            {
+                if (!project.FilePath.EndsWith($"{tagHelper.AssemblyName}.csproj", FilePathComparison.Instance))
+                {
+                    continue;
+                }
+
+                foreach (var path in project.DocumentFilePaths)
+                {
+                    // Rule out if not Razor component with correct name
+                    if (!IsPathCandidateForComponent(path, typeName))
+                    {
+                        continue;
+                    }
+
+                    // Get document and code document
+                    var documentSnapshot = project.GetDocument(path);
+                    var razorCodeDocument = await documentSnapshot.GetGeneratedOutputAsync().ConfigureAwait(false);
+                    if (razorCodeDocument is null)
+                    {
+                        continue;
+                    }
+
+                    // Make sure we have the right namespace of the fully qualified name
+                    if (!ComponentNamespaceMatchesFullyQualifiedName(razorCodeDocument, tagHelper.Name))
+                    {
+                        continue;
+                    }
+                    return documentSnapshot;
+                }
+            }
+            return null;
+        }
+
+        public static bool IsPathCandidateForComponent(string path, string componentName)
+        {
+            return path.EndsWith($"{componentName}.razor", FilePathComparison.Instance);
+        }
+
+        public static bool ComponentNamespaceMatchesFullyQualifiedName(RazorCodeDocument razorCodeDocument, string fullyQualifiedComponentName)
+        {
+            var namespaceNode = (NamespaceDeclarationIntermediateNode)razorCodeDocument
+                .GetDocumentIntermediateNode()
+                .FindDescendantNodes<IntermediateNode>()
+                .First(n => n is NamespaceDeclarationIntermediateNode);
+
+            DefaultRazorTagHelperBinderPhase.ComponentDirectiveVisitor.TrySplitNamespaceAndType(fullyQualifiedComponentName, out var namespaceNameSpan, out var typeNameSpan);
```

Comment:
For every single document you're extracting the namespace name span; however, can't that be done once and passed in?

Reply:
This is checking whether the namespace of the CodeDocument we're examining matches up with the namespace of the tag we're looking for. `<Component>` with tag helper `Namespace.Component` determines if the namespace declaration intermediate node has namespace name `Namespace`, which has to happen for every code document.

Reply:
But the callsite is: `if (!ComponentNamespaceMatchesFullyQualifiedName(razorCodeDocument, tagHelper.Name))` and `tagHelper.Name` corresponds to the `fullyQualifiedComponentName`. tagHelper.Name doesn't change. Am I missing something?

Reply:
OH. Oh oh oh. I am illiterate. Good call!